### ISSUE TITLE
systemd MemoryLimit is ignored and system's memory is used instead

### ISSUE THREAD

**Issue by fastest963** [cockroachdb/cockroach#31750]
**Describe the problem**

Running CockroachDB in systemd with:
```
MemoryLimit=6357M
```
Results in:
```
I181023 16:39:35.723845 1 server/status/recorder.go:609  available memory from cgroups (8.0 EiB) exceeds system memory 7.3 GiB, using system memory
I181023 16:39:35.724144 1 cli/start.go:1069  CockroachDB CCL v2.1.0-rc.1 (x86_64-unknown-linux-gnu, built 2018/10/22 22:39:56, go1.10.3)
I181023 16:39:35.825255 1 server/status/recorder.go:609  available memory from cgroups (8.0 EiB) exceeds system memory 7.3 GiB, using system memory
I181023 16:39:35.825300 1 server/config.go:385  system total memory: 7.3 GiB
```

This means that CockroachDB reports that it has 7.3GB available when it really only has 6.2GB available. `systemctl status cockroachdb` reports the limit:
```
sudo systemctl status cockroachdb
● cockroachdb.service - cockroachdb
   Loaded: loaded (/etc/systemd/system/cockroachdb.service; enabled; vendor preset: disabled)
   Active: active (running) since Tue 2018-10-23 16:39:36 UTC; 5min ago
 Main PID: 31550 (cockroach)
   Memory: 559.8M (limit: 6.2G)
   CGroup: /system.slice/cockroachdb.service
           └─31550 /usr/bin/cockroach start --logtostderr --no-color --certs-dir /etc/c...
```

**To Reproduce**

Create a systemd file like so:
```
# /etc/systemd/system/cockroachdb.service
[Unit]
After=network.target
Description=cockroachdb

[Service]
TimeoutStartSec=5
RestartForceExitStatus=SIGPIPE
CPUQuota=180%
StartLimitInterval=0
ExecStart=/usr/bin/cockroach \
  start \
  --logtostderr \
  --no-color
LimitNOFILE=50000
RestartSec=5
MemoryLimit=6357M
User=cockroachdb
TimeoutStopSec=60
Type=notify
Restart=always

[Install]
WantedBy=multi-user.target
```
And then observe that memory limit is reported incorrectly.

**Expected behavior**
I expected CockroachDB to understand that it's memory limit is 6.2GB not 7.3GB.

**Environment:**
 - CockroachDB version: v2.1.0-rc.1 (and v2.1.0-beta.20181015)
 - Server OS: CentOS 7.4

**Additional context**

Incorrect memory usage graphs and potentially CockroachDB trying to use more memory than is available and crashing.

**Comment by knz** [comment]
This needs to be investigated in our upstream `gosigar` package. `gosigar` is assuming the limits are properly configured via the standard OS API. It looks like systemd is doing something non-standard here (as it usually does), which will in turn require special support in `gosigar`.

I recommend filing the issue upstream (with a small `gosigar` repro that demonstrates the problem) and see what they say.

**Comment by fastest963** [comment]
I briefly looked into it and I guess that location isn't updated when a service has a memory limit. There's a way using `/proc/pid/cgroup` to get the cgroup and then use that to get the memory limit for that specific cgroup but I didn't fully investigate this yet. I'm out of town for the next few days but I'll do more digging when I get back. I was trying to figure out how other packages determine the memory limit to see if there was an easier way that I was missing.

**Comment by fastest963** [comment]
@bdarnell I tested https://github.com/fastest963/cockroach/commit/9e5ce85f457d72171ecea228c3e62da5126b0218 and it seems to work. Essentially it gets the cgroup name via `/proc/self/cgroup` and then gets the memory limit for that group, falling back to `/` (which is the current functionality).

**Comment by christianhuening** [comment]
+1 for this!

**Comment by tim-o** [comment]
Zendesk ticket #[3200](https://cockroachdb.zendesk.com/agent/tickets/3200) has been linked to this issue.

**Comment by bdarnell** [comment]
I believe this has been addressed by #43137 and #44549
